File: frontend/src/types/campaign.ts
import type { CampaignObjective, CampaignStatus } from './common'
import type { DerivedMetrics } from './metric'
import type { Platform } from './platform'

export interface Campaign {
  id: number
  name: string
  description: string | null
  platform: Platform
  objective: CampaignObjective
  budget: number
  startDate: string
  endDate: string | null
  status: CampaignStatus
  targetAudience: string | null
  ownerId: number | null
  ownerName: string | null
  notes: string | null
  externalCampaignId: string | null
  metrics: DerivedMetrics
  createdAt: string
  updatedAt: string
}

export interface CampaignRequest {
  name: string
  description?: string
  platformId: number
  objective: CampaignObjective
  budget: number
  startDate: string
  endDate?: string | null
  targetAudience?: string
  ownerId?: number | null
  notes?: string
  externalCampaignId?: string
}

File: backend/src/main/java/com/marketinganalytics/platform/entity/Campaign.java
package com.marketinganalytics.platform.entity;

import com.marketinganalytics.platform.entity.enums.CampaignObjective;
import com.marketinganalytics.platform.entity.enums.CampaignStatus;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;

@Entity
@Table(name = "campaigns")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Campaign {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 150)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "platform_id", nullable = false)
    private Platform platform;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private CampaignObjective objective;

    @Column(nullable = false, precision = 14, scale = 2)
    private BigDecimal budget;

    @Column(name = "start_date", nullable = false)
    private LocalDate startDate;

    @Column(name = "end_date")
    private LocalDate endDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private CampaignStatus status = CampaignStatus.DRAFT;

    @Column(name = "target_audience", columnDefinition = "TEXT")
    private String targetAudience;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id")
    private User owner;

    @Column(columnDefinition = "TEXT")
    private String notes;

    /** ID of this campaign in the connected ad platform's own system, used to sync real metrics. */
    @Column(name = "external_campaign_id", length = 120)
    private String externalCampaignId;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @PrePersist
    void onCreate() {
        Instant now = Instant.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = Instant.now();
    }
}
